collectors/monitoring_collector.go

```diff
@@ -310,18 +224,47 @@ func (c *MonitoringCollector) reportMonitoringMetrics(ch chan<- prometheus.Metri
 		go func(metricsTypePrefix string) {
 			defer wg.Done()
 			level.Debug(c.logger).Log("msg", "listing Google Stackdriver Monitoring metric descriptors starting with", "prefix", metricsTypePrefix)
-			ctx := context.Background()
 			filter := fmt.Sprintf("metric.type = starts_with(\"%s\")", metricsTypePrefix)
 			if c.monitoringDropDelegatedProjects {
 				filter = fmt.Sprintf(
 					"project = \"%s\" AND metric.type = starts_with(\"%s\")",
 					c.projectID,
 					metricsTypePrefix)
 			}
-			if err := c.monitoringService.Projects.MetricDescriptors.List(utils.ProjectResource(c.projectID)).
-				Filter(filter).
-				Pages(ctx, metricDescriptorsFunction); err != nil {
-				errChannel <- err
+
+			req := &monitoringpb.ListMetricDescriptorsRequest{
+				Name:   utils.ProjectResource(c.projectID),
+				Filter: filter,
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+			defer cancel()
+			it := c.metricClient.ListMetricDescriptors(ctx, req)
+
+			apiCalls := 1.0
+			var descriptors []*metric.MetricDescriptor
+			for {
+				// There's nothing exposed in https://pkg.go.dev/google.golang.org/api/iterator@v0.103.0 which lets you
+				// know an API call was initiated. You might think https://pkg.go.dev/google.golang.org/api/iterator@v0.103.0#NewPager
+				// could do it but the pageSize is a superficial page size that the consumer sets and has impact on API call paging.
+				// If we know there are no items left in the current page and there's a non-empty page token then calling Next() is going to initiate an API call
+				if it.PageInfo().Remaining() == 0 && it.PageInfo().Token != "" {
+					apiCalls += 1.0
+				}
+				descriptor, err := it.Next()
+				if err == iterator.Done {
+					break
+				}
+				if err != nil {
+					errChannel <- fmt.Errorf("error while fetching descriptors for %s, %v", metricsTypePrefix, err)
+					break
+				}
+				descriptors = append(descriptors, descriptor)
+			}
+			c.apiCallsTotalMetric.Add(apiCalls)
+
+			if err := c.collectMetricsForDescriptors(ch, begun, descriptors); err != nil {
+				errChannel <- fmt.Errorf("error while fetching descriptors for %s, %v", metricsTypePrefix, err)
```

Comment:
This is a behavioral difference as the previous implementation submitted each page to collect metrics. There's no documented default page size on the MetricDescriptor API but I don't think this ever resulted in multiple pages.